File: server/docker/runner.cpp
#include <iostream>
#include <cstdlib>
#include <string>
#include <unistd.h>
#include <sys/wait.h>

int main() {
    // Check if main.cpp exists
    if (access("/app/main.cpp", F_OK) != 0) {
        std::cerr << "Error: /app/main.cpp not found\n";
        return 2;
    }

    // Compile the code
    int compile_result = system("g++ -o /app/program /app/main.cpp");
    if (compile_result != 0) {
        std::cerr << "Compilation failed with exit code " << compile_result << "\n";
        return 1;
    }

    // Execute the compiled program
    execl("/app/program", "program", nullptr);
    
    // If execl returns, there was an error
    perror("Failed to execute program");
    return 1;
}

File: server/docker/runner.py
import sys
import traceback
from pathlib import Path

def run_code():
    try:
        # Check if main.py exists
        main_file = Path('/app/main.py')
        if not main_file.exists():
            print("Error: /app/main.py not found", file=sys.stderr)
            sys.exit(2)
            
        # Read and execute the code
        with open(main_file, 'r') as f:
            code = f.read()
            
        # Execute in a clean namespace
        exec(code, {'__name__': '__main__'})
            
    except Exception as e:
        print(f"Error: {str(e)}", file=sys.stderr)
        print(traceback.format_exc(), file=sys.stderr)
        sys.exit(1)

if __name__ == '__main__':
    run_code()